feat(cms): import/export of content

allardy opened the pull request:
Still require extensive testing to ensure it handles correctly a lot of elements. Missing validation of import file.

- Export/Import content from the CMS (similar to QNA)
- Ensures correct content types are present (they are not exported to the file since they are global at the moment and couldn't be safely imported)
- Fixed an issue where the number of elements in content types was not updated when content was added or removed

![image](https://user-images.githubusercontent.com/42552874/67834826-cb1e4680-fabe-11e9-9e1a-dbf882445fc6.png)

![image](https://user-images.githubusercontent.com/42552874/67834969-27816600-fabf-11e9-9dda-d7cc1c7c7197.png)

allardy committed:
feat(cms): import/export of content

allardy committed:
ajustments

allardy committed:
Merge branch 'patch/12-2-3' into ya-cms-export

allardy committed:
Merge branch 'patch/12-2-3' into ya-cms-export

EFF commented on src/bp/core/routers/bots/content.ts:
```diff
@@ -129,6 +134,70 @@ export class ContentRouter extends CustomRouter {
+      this._checkTokenHeader,
+      this._needPermissions('read', 'bot.content'),
+      async (req, res) => {
+        const elements = await this.cms.getAllElements(req.params.botId)
```
what if there's a really big load of elements ? should we break down this function un chunks ?

allardy replied:
I think there's a couple of places we could hit that issue, but it doesn't seems to be a problem at the moment. I'll leave a todo for future improvement

EFF commented on src/bp/core/routers/bots/content.ts:
```diff
@@ -129,6 +134,70 @@ export class ContentRouter extends CustomRouter {
+          res.send({
+            cmsCount: (existingElements && existingElements.length) || 0,
+            fileCmsCount: (importData && importData.length) || 0,
+            missingContentTypes: _.difference(importedContentTypes, contentTypes)
```
missing, is missing content elements to override all elements ?

allardy replied:
This makes sure that every content types used by the imported elements are present, otherwise they must be added beforehand.

EFF commented on src/bp/core/routers/bots/content.ts:
```diff
@@ -129,6 +134,70 @@ export class ContentRouter extends CustomRouter {
+          const importData: ContentElement[] = await JSON.parse(req.file.buffer)
+
+          for (const { contentType, formData, id } of importData) {
+            await this.cms.createOrUpdateContentElement(req.params.botId, contentType, formData, id)
```
even if not cleaar_insert, this will overwrite all elements ? 

is clear_insert only to delete missing content elements in inport data ?

allardy replied:
Yep, the message displayed to the user is more clear when making a choice. You can either create new elements and update existing ones, or clear all of the existing ones, then create them

EFF reviewed (approved):
Although this probably works fine, JSON doesn't make life easy for translators .. I know we probably discussed it already but my opinion is CSV would be a better option for translators.

allardy commented:
CSV files are the devil, as we saw with the QNA module... That's the reason we moved to JSON and dropped CSV support. 

I think it would be a lot more useful to have a dedicated interface for translating, anything better than a csv

allardy committed:
pr comment
